File: trunk/daetools-package/daetools/dae_simulator/web_service.py
"""********************************************************************************
                            web_service.py
                 DAE Tools: pyDAE module, www.daetools.com
                 Copyright (C) Dragan Nikolic
***********************************************************************************
DAE Tools is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License version 3 as published by the Free Software
Foundation. DAE Tools is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with the
DAE Tools software; if not, see <http://www.gnu.org/licenses/>.
********************************************************************************"""
import os, sys, json, traceback, cgi
import uuid, zipfile, cgitb, hashlib, threading
import wsgiref.simple_server
from daetools.pyDAE import *

try:
    # python 3
    import http.client as http_client
    from urllib.parse import urlparse
except ImportError:
    # python 2
    import httplib as http_client
    from urlparse import urlparse

# cgi traceback
logdir = os.path.join(os.path.expanduser('~'), 'daetools_ws')
cgitb.enable(display = 0, logdir = logdir, format = 'html') 

class daeWebService(object):
    # The base class for web applications/web services.
    # Performs the low level tasks such as parsing the query to get its arguments
    # and sending responses with results to the clients.
    def __init__(self):
        self.noHTTPRequests = 0
        self.activeObjects = {}
        self.httpServer    = None
    
    @staticmethod
    def startWebService(application, address, port, serveForever = False):
        # Start the web application.
        http_port    = int(port)
        http_address = str(address)
        
        httpd = wsgiref.simple_server.make_server(http_address, http_port, application, handler_class=NoLoggingWSGIRequestHandler)
        application.httpServer = httpd
        if serveForever:
            # Serve even if there is no active clients.
            httpd.serve_forever(poll_interval = 5.0)
        else:
            # Serve as long as the function keepRunning returns True.
            # keepRunning will return False if all clients freed their objects and exited.
            while application.keepRunning():
                httpd.handle_request()

    @staticmethod
    def tryConnectWebService(address, port, timeout = 10):
        # Try to connect to the above server (timeout is 10s).
        # If successful, the server has successfully been started
        # and the subsequent clients can connect to it.
        c = http_client.HTTPConnection(address, port, timeout)
    
    def keepRunning(self):
        # Server will stop if at least one request has been processed and no activeObjects exist
        # (all active objects finished with the simulation).
        # Therefore, the server can serve multiple clients but will shutdown after all objects 
        # freed their resources and the number of active objects goes to zero.
        return len(self.activeObjects) > 0 or self.noHTTPRequests == 0
    
    def getQueryArguments(self, environ):
        # Returns the query arguments' values as a dictionary object.
        # Can process both GET and POST requests (double check this).
        arguments   = cgi.FieldStorage(fp = environ['wsgi.input'], environ = environ) 
        pathInfo    = environ['PATH_INFO']
        return pathInfo, arguments
        
    def jsonError(self, reason, start_response, simulationID):
        # Returns the error status to the client.
        # The argument reason is a string with the exception description.
        # In case of errors always deletes the current object identified by its simulationID.
        if simulationID and simulationID in self.activeObjects:
            del self.activeObjects[simulationID]
        obj = {'Status': 'Error',
               'Reason': reason,
               'Result': None}
        return self._sendResponse_json(obj, '500 Internal Server Error', start_response)
        
    def jsonSuccess(self, start_response):
        # Returns the success status to the client and no results.
        obj = {'Status': 'Success',
               'Reason': None,
               'Result': None}
        return self._sendResponse_json(obj, '200 OK', start_response)

    def jsonResult(self, result, start_response):
        # Returns the success status and the data to the client.
        obj = {'Status': 'Success',
               'Reason': None,
               'Result': result}
        return self._sendResponse_json(obj, '200 OK', start_response)
    
    def _sendResponse_json(self, obj, status, start_response):
        # Low level function that sends the HTTP response in JSON format.
        # The argument obj contains the data to be sent as JSON message.
        json_b = json.dumps(obj, indent = 2).encode() # default is utf-8
        start_response(status, [('Access-Control-Allow-Origin', '*'),
                                ('Access-Control-Allow-Methods', 'POST, GET, OPTIONS'),
                                ('Content-type', 'application/json'),
                                ('Content-Length', str(len(json_b)))])
        return [json_b]

    def _sendResponse_html(self, html_content, status, start_response):
        # Low level function that sends the HTTP response in HTML format.
        # The argument html_content contains the HTML message.
        html_b = html_content.encode() # default is utf-8
        start_response(status, [('Access-Control-Allow-Origin', '*'),
                                ('Access-Control-Allow-Methods', 'POST, GET, OPTIONS'),
                                ('Content-type', 'text/html'),
                                ('Content-Length', str(len(html_b)))])
        return [html_b]

    def _sendFile(self, filePath, contentType, start_response):
        # Low level function that sends a file as the HTTP response.
        # The argument html_content contains the HTML message.
        try:
            f = open(filePath, 'rb')
            content     = f.read()
            filename    = os.path.basename(filePath)
            content_len = len(content)
            start_response('200 OK', [('Access-Control-Allow-Origin', '*'),
                                      ('Access-Control-Allow-Methods', 'POST, GET, OPTIONS'),
                                      ('Content-type', contentType),
                                      ('Content-Disposition', 'attachment; filename=%s' % filename),
                                      ('Content-Length', str(content_len))])
        finally:
            f.close()
        return [content]

class NoLoggingWSGIRequestHandler(wsgiref.simple_server.WSGIRequestHandler):
    # Request handler that suppresses logging to the stderr.
    def log_message(self, format, *args):
        pass

File: trunk/daetools-package/daetools/dae_simulator/javascript/web_service.js
/***********************************************************************************
                           web_service.js
                 DAE Tools Project, www.daetools.com
                 Copyright (C) Dragan Nikolic
************************************************************************************
DAE Tools is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License version 3 as published by the Free Software
Foundation. DAE Tools is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with the
DAE Tools software; if not, see <http://www.gnu.org/licenses/>.
************************************************************************************/
class daeWebService
{
    constructor(address, port, webServiceName, method)
    {
        this.address        = address;
        this.port           = port;
        this.webServiceName = webServiceName;
        this.method         = method;
        this.asyncRequests  = false;
    }
    
    get ServerStatus()
    {
        var parameters = {};
        var functionName = 'status';
        var response = this.executeFun(functionName, parameters);
        if(response.success == false)
            throw response['Reason'];
        return response.result['ActiveClients'];
    }

    ClearServer()
    {
        var parameters = {};
        var functionName = 'clear';
        var response = this.executeFun(functionName, parameters);
        if(response.success == false)
            throw response['Reason'];
    }
    
    onSuccess(httpRequest, path, args) 
    {
    }
    
    onError(httpRequest, path, args) 
    {
        var msg  = 'Request failed';
        var msg1 = 'Query: ' + path + '?' + args.toString();
        var msg2 = 'Status: ' + httpRequest.status.toString() + ' ' + httpRequest.statusText;
        var msg3 = 'Headers: ' + httpRequest.getAllResponseHeaders();
        var msg4 = 'Response: \n' + httpRequest.responseText;
        console.log(msg);
        console.log(msg1);
        console.log(msg2);
        console.log(msg3);
        console.log(msg4);
        alert(msg + '\n' + msg1 + '\n' + msg2 + '\n' + msg3 + '\n' + msg4 + '\n\n');
    }
    
    onConnectionFailure(path, error)
    {
        var msg = '';
        if(error !== null)
            msg = error;
        alert(error + '\n\n' + 
              'Connection failure to the server: ' + 
              path + '\n' +
              'The web service has not been started.\nStart it using: python -m daetools.dae_simulator.' + this.webServiceName);
    }
    
    getResponse(httpRequest) 
    {
        var res = {success:false, 
                   reason: "", 
                   result: null};
                
        if (httpRequest.readyState == 4)
        {
            if(httpRequest.status == 200) 
            {
                // Success: '200 OK'.
                var response = JSON.parse(httpRequest.responseText);
                if(response["Status"] == "Success")
                    res.success = true;
                res.result = response["Result"];
            }
            else if(httpRequest.status == 400)
            {
                // Error in client request: '400 BAD REQUEST'.
                var response = JSON.parse(httpRequest.responseText);
                res.reason = response["Reason"];
            }
            else if(httpRequest.status == 500)
            {
                // Server error: '500 INTERNAL SERVER ERROR'.
                var response = JSON.parse(httpRequest.responseText);
                res.reason = response["Reason"];
            }
            else
            {
                res.reason = "Unknown status received: " + httpRequest.status.toString() + " " + httpRequest.statusText;
            }
        } 
        
        return res;
    }
    
    createHTTPRequest()
    {
        if (window.XMLHttpRequest)
            return new XMLHttpRequest();
        else
            return new ActiveXObject("Microsoft.XMLHTTP");
    };
    
    executeFun(functionName, parameters) 
    {
        var webApp = this;
        var httpRequest = this.createHTTPRequest();
        var args = 'function=' + functionName;
        for(var paramName in parameters) 
            args += '&' + String(paramName) + '=' + String(parameters[paramName]);
        var path = this.address + ':' + this.port.toString() + '/' + this.webServiceName;
        httpRequest.open(this.method, path, this.asyncRequests);
        httpRequest.setRequestHeader('Accept', 'application/json');
        httpRequest.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
        httpRequest.addEventListener('load', function(text) { console.log(text); });
        httpRequest.onreadystatechange = function() {
            // This function is used for asynchronous request
            if(this.readyState == 4)
            {
                if(httpRequest.status == 200) 
                {
                    // Success: '200 OK'.
                    var response = webApp.getResponse(httpRequest);
                    if(response.success == true)
                        webApp.onSuccess(httpRequest, path, args);
                    else
                        webApp.onError(httpRequest, path, args);
                }
                else if(httpRequest.status == 400)
                {
                    // Error in client request: '400 BAD REQUEST'.
                    webApp.onError(httpRequest, path, args);
                }
                else if(httpRequest.status == 500)
                {
                    // Server error: '500 INTERNAL SERVER ERROR'.
                    webApp.onError(httpRequest, path, args);
                }
                else if(httpRequest.status == 0)
                {
                    // Connection failure
                }
                else
                {
                    webApp.onError(httpRequest, path, args);
                }
            }
        };
        
        /* Send the request with the parameters uri encoded. */
        try
        {
            httpRequest.send(encodeURI(args));
        }
        catch(exception)
        {
            webApp.onConnectionFailure(path, exception);
            throw exception;
        }
        
        /* If the request sent was synchronous then return the response.
           Otherwise, return nothing. */
        if(!this.asyncRequests)
        {
            var response = this.getResponse(httpRequest);
            if(response.success == false)
                throw response['Reason'];
            return response;
        }
        else
            return null;
    }
}

/* Gets the server status.
   This function is run using the setTimeout (delay = 0) to improve responsiveness. */               
function getServerStatus(ws)
{
    var getServerStatus_ = function() {
        if(ws == null)
            return;
        
        var path   = ws.address + ':' + ws.port.toString() + '/' + ws.webServiceName;
        var status = ws.ServerStatus;
        var msg = 'Web service: ' + path + '\n';
        msg += 'Number of active clients: ' + status.length;
        alert(msg);
    };
    setTimeout(getServerStatus_, 0);
}

/* Clears the server (removes all clients; use with care).
   This function is run using the setTimeout (delay = 0) to improve responsiveness. */               
function clearServer(ws)
{
    var clearServer_ = function() {
        if(ws == null)
            return;
        
        ws.ClearServer();
        clearInputs();
        simulation = null;
    };
    setTimeout(clearServer_, 0);
}
